sdks/python/apache_beam/io/mongodbio_test.py

```diff
@@ -0,0 +1,187 @@
+
+    result = []
+    for i in self.mongo_source.read(mock_tracker):
+      result.append(i)
+    self.assertListEqual([{'x': 1}, {'x': 2}], result)
+
+  def test_display_data(self):
+    data = self.mongo_source.display_data()
+    self.assertTrue('uri' in data)
+    self.assertTrue('database' in data)
+    self.assertTrue('collection' in data)
+
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test__get_avg_document_size(self, mock_client):
+    mock_client.return_value.__enter__.return_value.__getitem__ \
+      .return_value.command.return_value = {'avgObjSize': 5}
+    self.assertEqual(5, self.mongo_source._get_avg_document_size())
+
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test_get_document_count(self, mock_client):
+    mock_client.return_value.__enter__.return_value.__getitem__ \
+      .return_value.__getitem__.return_value.count_documents.return_value = 10
+
+    self.assertEqual(10, self.mongo_source._get_document_count())
+
+
+class ReadFromMongoDBTest(unittest.TestCase):
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test_read_from_mongodb(self, mock_client):
+    objects = [{'x': 1}, {'x': 2}]
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value. \
+      command.return_value = {'avgObjSize': 1}
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value. \
+      __getitem__.return_value.find.return_value = objects
+    mock_client.return_value.__enter__.return_value.__getitem__.return_value. \
+      __getitem__.return_value.count_documents.return_value = 2
+
+    with TestPipeline() as p:
+      docs = p | 'ReadFromMongoDB' >> ReadFromMongoDB(
+          uri='mongodb://test', db='db', coll='collection')
+      assert_that(docs, equal_to(objects))
+
+
+class GenerateObjectIdFnTest(unittest.TestCase):
+  def test_process(self):
+    with TestPipeline() as p:
+      output = (p | "Create" >> beam.Create([{
+          'x': 1
+      }, {
+          'x': 2,
+          '_id': 123
+      }])
+                | "Generate ID" >> beam.ParDo(_GenerateObjectIdFn())
+                | "Check" >> beam.Map(lambda x: '_id' in x))
+      assert_that(output, equal_to([True] * 2))
+
+
+class WriteMongoFnTest(unittest.TestCase):
+  @mock.patch('apache_beam.io.mongodbio._MongoSink')
+  def test_process(self, mock_sink):
+    docs = [{'x': 1}, {'x': 2}, {'x': 3}]
+    with TestPipeline() as p:
+      _ = (p | "Create" >> beam.Create(docs)
+           | "Write" >> beam.ParDo(_WriteMongoFn(batch_size=2)))
+      p.run()
+
+      self.assertEqual(
+          2, mock_sink.return_value.__enter__.return_value.write.call_count)
+
+  def test_display_data(self):
+    data = _WriteMongoFn(batch_size=10).display_data()
+    self.assertEqual(10, data['batch_size'])
+
+
+class MongoSinkTest(unittest.TestCase):
+  @mock.patch('apache_beam.io.mongodbio.MongoClient')
+  def test_write(self, mock_client):
```

Comment:
Also add a tests to confirm that (1)we accept unique IDs that are already available if set (2) if unique IDs are not set we generate them before writing.

Reply:
Added the cases in WriteToMongoDBTest.